### PR TITLE
Overhaul Field with new types, use Equinox, and introduce interfaces

### PR DESCRIPTION
**Pull request by diptodip**
This is a major set of changes that completes the exploration in Draft PR #157 and fixes issue #155 and issue #172. 

Flax/Linen is now removed from the entire library. All elements as well as the Fields themselves are now using Equinox, and Fields have no static attributes. Various new interfaces are introduced via abstract Equinox `Modules`, e.g. `Spectrum`, `Sensor`, `Sample`, `Monochromatic`, etc.

The Field is now split into four types: `ScalarField`, `ChromaticScalarField`, `VectorField`, and `ChromaticVectorField`. It is possible we will have new types of Fields in the future, which will compose a different set of interfaces.

For those using the functional API, the only change required for most people (using monochromatic simulations) should be that the spectral density arguments are no longer present. You can just pass a single float for the spectrum without specifying any density. For chromatic `Field`s, see `Spectrum` documentation. Also, shapes are much less complicated especially in the case of monochromatic scalar diffraction (which uses `ScalarField`). No extraneous empty dimensions are required anymore thanks to the different `Field` types.

Along with these changes, the documentation and examples have been totally overhauled with more realistic examples and clearer shape annotations.

While this PR introduces some breaking changes, they are relatively small and most people should only have to change a single line to upgrade.

### CODE REVIEW

**Comment by diptodip** [comment]
I should note that we will likely merge this soon regardless of any issues we find right now, though there are a few more spots in the documentation that are still missing, functions that do not yet have corresponding elements, etc.

File: pyproject.toml [hunk 1]
```diff
@@ -17,20 +17,21 @@ dependencies = [
...
+  "jaxtyping>=0.3.3",
   "equinox>=0.12.2",
   "optimistix>=0.0.10",
 ]
-version = "0.4.0"
+version = "0.5.0"
 
 [project.optional-dependencies]
-dev = ["mypy>=0.991", "pytest>=7.2.0", "ruff==0.9.0", "pre-commit>=4.0.0", "matplotlib>=3.10.0"]
+dev = ["mypy>=0.991", "pytest>=7.2.0", "ruff==0.11.12", "pre-commit>=4.0.0", "matplotlib>=3.10.0"]
```
Comment (GJBoth) [inline pyproject.toml]: Ruff is already on 0.14, but we can fix that later. Also have to make sure this aligns with the ruff version in the CI.

File: examples/parallel_imaging.py [hunk 2]
```diff
@@ -5,106 +5,110 @@
...
 
 
 volume = jnp.ones(
-    (num_devices, num_planes_per_device, *shape, 1, 1)
+    (num_devices, num_planes_per_device, *shape)
 )  # fill in your volume here
-z = jnp.linspace(-4, 4, num=num_planes).reshape(num_devices, num_planes_per_device)
-params = init_params(jax.random.split(jax.random.PRNGKey(6022), num_devices), volume, z)
-widefield_image = compute_image(params, volume, z)
+volume = jax.device_put_sharded(
```
Comment (GJBoth) [inline examples/parallel_imaging.py]: `pmap` above shouldn't be necessary when using sharded, right?

File: docs/FAQ.md [hunk 3]
```diff
@@ -10,38 +10,9 @@ Chromatix tries to respect composable `jax` transformations, so you can use all
...
-```
-
-We discuss `trainable()` further in the [API documentation](api/utils.md#chromatix.utils.utils.trainable).
 
 ## How can I optimize parameters in `chromatix`?
 
-Everything in Chromatix is just a PyTree or `jax`-transformable function, so you can use any `jax` optimization program or library you like! For deep learning style optimizers we suggest having a look at [optax](https://github.com/deepmind/optax). For classical optimization schemes, including optimizers found in `scipy` and conjugate gradient methods (but all implicitly differentiable!) see [`jaxopt`](https://github.com/google/jaxopt).
+Everything in Chromatix is just a PyTree or `jax`-transformable function, so you can use any `jax` optimization program or library you like! For deep learning style optimizers we suggest having a look at [optax](https://github.com/deepmind/optax). For classical optimization schemes, including optimizers found in `scipy` and conjugate gradient methods (but all implicitly differentiable!) see [`jaxopt`](https://github.com/google/jaxopt). Also see our [documentation on optimization](https://chromatix.readthedocs.io/en/latest/training/).
```
Comment (GJBoth) [inline docs/FAQ.md]: Jaxopt isn't developed anymore... should we link to optimistix?

File: src/chromatix/core/base.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    @abc.abstractmethod
+    def __call__(self, sensor_input: Field | Array) -> Array:
+        pass
+
+    @abc.abstractmethod
+    def resample(self, resample_input: Array, input_spacing: ScalarLike) -> Array:
+        pass
+
+
+class Resampler(eqx.Module, strict=_strict_config):
```
Comment (GJBoth) [inline src/chromatix/core/base.py]: Whats the difference between Resampler and Sensor here?

File: src/chromatix/data/data.py
Comment (GJBoth) [inline src/chromatix/data/data.py]: I think we need to clean this up sometime... but not necessary for now.

File: src/chromatix/elements/samples.py [hunk 5]
```diff
@@ -1,24 +1,95 @@
...
+
+class ClearThinSample(Sample, Scattering, strict=True):
+    """
+    Perturbs an incoming ``ScalarField`` as if it went through a clear
+    thin sample object with a given refractive index change ``dn`` from the
+    refractive index of the medium and of a given ``thickness`` in the same
+    units as the spectrum of the incoming ``ScalarField``.
 
     The sample is supposed to follow the thin sample approximation, so the
     sample perturbation is calculated as:
```
Comment (GJBoth) [inline src/chromatix/elements/samples.py]: Remove absorption here.

File: docs/training.ipynb
[the remark this answers is not in the document]
Comment (diptodip) [reply docs/training.ipynb]: It doesn't actually block vmapping on the call which is what I imagine people would usually want to vmap over in that style. The thing about the `partition` approach is that you have to create this `filter_spec` tree yourself (as you see in that notebook) which I think is just as if not more confusing than the Flax/Linen `init`/`apply` approach. So I think it's fine if those classes are just holding a parameter plus some static metadata.

Also to answer your question from above, yeah setting some things as static is mostly about getting things to work with the default JAX transforms. It's usually on stuff like integer shapes or boolean flags. I also figure if this update is about convenience, then the stuff we have for convenience should really just handle the default cases well.

File: src/chromatix/core/base.py (hunk 4 above)
Comment (diptodip) [reply src/chromatix/core/base.py]: The `Resampler` is just a way to create an interface for the `resample_fn` that we had for `Sensor`s before. It only does resampling on arrays, no sensor stuff. We could now think about using this approach (that supports cubic interpolation) in other places that need interpolation as well. Also this way it should be more clear how to construct them if you want to use the functional API or develop your own `Sensor`. But I also note in the docstrings that you could just use `BasicSensor` most of the time, which is now really easy because you don't have to handle the state initialization yourself.

**Comment by diptodip** [comment]
> very bullish on this PR. I think it'll fix 90% of the issues we've been having - and most of my comments are pretty minor. I see no blocking issues, so we could go ahead and merge this into main and fix the minor things as they come up? Few general comments:
> 
> * We currently use `eqx.field(static=True)` quite a bit - is that so we can use `jax.jit` instead of `eqx.filter_jit`?
> * I think everything is fully typed now, should we add type checking to the CI, and turn on beartype as well since we have shape annotations?

I'm against runtime type checking or even type checking in the CI for now. I really look at the type annotations as hints for people reading the function signature to get their inputs correct, which means that these type hints can be kind of fuzzy but still useful. I think there are just going to be too many corner cases with that sort of thing for a language like Python (even though we're basically writing compile time code for JAX) that I think are not worth the small benefits. But maybe we can try it out and see later!
